cache/cache.go

```diff
@@ -0,0 +1,186 @@
+package cache
+
+import (
+	"bytes"
+	"sort"
+	"sync"
+
+	"go.uber.org/atomic"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+)
+
+type ATXData struct {
+	Weight             uint64
+	BaseHeight, Height uint64
+	Nonce              types.VRFPostIndex
+	Malicious          bool
+}
+
+type Opt func(*Cache)
+
+// WithCapacity sets the number of epochs from the latest applied
+// that cache will maintain in memory.
+func WithCapacity(capacity types.EpochID) Opt {
+	return func(cache *Cache) {
+		cache.capacity = capacity
+	}
+}
+
+func New(opts ...Opt) *Cache {
+	cache := &Cache{
+		capacity: 2,
+		epochs:   map[types.EpochID]epochCache{},
+	}
+	for _, opt := range opts {
+		opt(cache)
+	}
+	return cache
+}
+
+type Cache struct {
+	evicted atomic.Uint32
+
+	// number of epochs to keep
+	// capacity is enforced by the cache itself
+	capacity types.EpochID
+
+	mu     sync.RWMutex
+	epochs map[types.EpochID]epochCache
+}
+
+type epochCache struct {
+	atxs       map[types.ATXID]*ATXData
+	identities map[types.NodeID][]types.ATXID
+}
+
+func (c *Cache) Evicted() types.EpochID {
+	return types.EpochID(c.evicted.Load())
+}
+
+func (c *Cache) IsEvicted(epoch types.EpochID) bool {
+	return c.evicted.Load() >= epoch.Uint32()
+}
+
+// OnApplied is a notification for cache to evict epochs that are not useful
+// to keep in memory.
+func (c *Cache) OnApplied(applied types.EpochID) {
+	if applied < c.capacity {
+		return
+	}
+	evict := applied - c.capacity
+	if c.IsEvicted(evict) {
+		return
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.evicted.Load() < applied.Uint32() {
+		c.evicted.Store(evict.Uint32())
+	}
+	for epoch := range c.epochs {
+		if epoch <= evict {
+			delete(c.epochs, epoch)
+			atxsCounter.DeleteLabelValues(epoch.String())
+			identitiesCounter.DeleteLabelValues(epoch.String())
+		}
+	}
+}
+
+func (c *Cache) Add(epoch types.EpochID, node types.NodeID, atx types.ATXID, data *ATXData) {
+	if c.IsEvicted(epoch) {
+		return
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	ecache, exists := c.epochs[epoch]
+	if !exists {
+		ecache = epochCache{
+			atxs:       map[types.ATXID]*ATXData{},
+			identities: map[types.NodeID][]types.ATXID{},
+		}
+		c.epochs[epoch] = ecache
+	}
+	if _, exists := ecache.atxs[atx]; exists {
+		return
+	}
+	ecache.atxs[atx] = data
+	atxs := ecache.identities[node]
+
+	atxsCounter.WithLabelValues(epoch.String()).Add(1)
+	if atxs == nil {
+		identitiesCounter.WithLabelValues(epoch.String()).Add(1)
+	}
+
+	atxs = append(atxs, atx)
+	// NOTE(dshulyak) doesn't make sense, as we don't guarantee that every node
+	// will see same atxs. it actually should see atmost one and malfeasence proof if node equivocated.
+	// find out if we have use case in consensus.
```

Comment:
i don't think we need this. the identity should already be set malicious when Cache.SetMalicious() is called.
and at that point it shouldn't matter which ATX is returned for a malicious identity.

Reply:
we need to store them to support NodeHasAtx. otherwise ballots won't be accepted. and for consistency it is better to keep them ordered

Reply:
there was a plan to only store the first equivocating atx/ballot and only sync those when they are referenced.
i see NodeHasAtx is used to check proposal eligibility. do you think this step is needed if the ballot is malicious?

Reply:
why does it matter?